internal/cmd/update.go

```diff
@@ -0,0 +1,309 @@
+		}
+	}
+
+	desiredVersion, err := getAPIVersionUnauthed(u.httpClient, *coderURL)
+	if err != nil {
+		return clog.Fatal("fetch api version", clog.Causef(err.Error()))
+	}
+
+	clog.LogInfo(fmt.Sprintf("Coder instance at %q reports version %s", coderURL.String(), desiredVersion.String()))
+	clog.LogInfo(fmt.Sprintf("Current version of coder-cli is %s", version.Version))
+
+	if currentVersion, err := semver.Make(u.versionF()); err == nil {
+		if desiredVersion.Compare(currentVersion) == 0 {
+			clog.LogInfo("Up to date!")
+			return nil
+		}
+	}
+
+	if !force {
+		label := fmt.Sprintf("Update coder-cli to version %s", desiredVersion.FinalizeVersion())
+		if _, err := u.confirmF(label); err != nil {
+			return clog.Fatal("failed to confirm update", clog.Tipf(`use "--force" to update without confirmation`))
+		}
+	}
+
+	downloadURL := makeDownloadURL(desiredVersion.FinalizeVersion(), runtime.GOOS, runtime.GOARCH)
+
+	var downloadBuf bytes.Buffer
+	memWriter := bufio.NewWriter(&downloadBuf)
+
+	clog.LogInfo("fetching coder-cli from GitHub releases", downloadURL)
+	resp, err := u.httpClient.Get(downloadURL)
+	if err != nil {
+		return clog.Fatal(fmt.Sprintf("failed to fetch URL %s", downloadURL), clog.Causef(err.Error()))
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return clog.Fatal("failed to fetch release", clog.Causef("URL %s returned status code %d", downloadURL, resp.StatusCode))
+	}
+
+	if _, err := io.Copy(memWriter, resp.Body); err != nil {
+		return clog.Fatal(fmt.Sprintf("failed to download %s", downloadURL), clog.Causef(err.Error()))
+	}
+
+	_ = resp.Body.Close()
+
+	if err := memWriter.Flush(); err != nil {
+		return clog.Fatal(fmt.Sprintf("failed to save %s", downloadURL), clog.Causef(err.Error()))
+	}
+
+	// TODO: validate the checksum of the downloaded file. GitHub does not currently provide this information
+	// and we do not generate them yet.
+	updatedBinary, err := extractFromArchive("coder", downloadBuf.Bytes())
+	if err != nil {
+		return clog.Fatal("failed to extract coder binary from archive", clog.Causef(err.Error()))
+	}
+
+	// We assume the binary is named coder and write it to coder.new
+	updatedCoderBinaryPath := u.executablePath + ".new"
+	updatedBin, err := u.fs.OpenFile(updatedCoderBinaryPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, currentBinaryStat.Mode().Perm())
+	if err != nil {
+		return clog.Fatal("failed to create file for updated coder binary", clog.Causef(err.Error()))
+	}
+
+	fsWriter := bufio.NewWriter(updatedBin)
+	if _, err := io.Copy(fsWriter, bytes.NewReader(updatedBinary)); err != nil {
+		return clog.Fatal("failed to write updated coder binary to disk", clog.Causef(err.Error()))
+	}
+
+	if err := fsWriter.Flush(); err != nil {
+		return clog.Fatal("failed to persist updated coder binary to disk", clog.Causef(err.Error()))
+	}
+
+	if err = u.fs.Rename(updatedCoderBinaryPath, u.executablePath); err != nil {
+		return clog.Fatal("failed to update coder binary in-place", clog.Causef(err.Error()))
+	}
```

Comment:
Can we test that we can execute the new binary before we rename?

Reply:
Done